picard/metadata.py

```diff
@@ -199,6 +200,27 @@ def compare_to_release(self, release, weights):
         sim = linear_combination_of_weights(parts) * get_score(release)
         return SimMatchRelease(similarity=sim, release=release)
 
+    @staticmethod
+    def extract_year_from_date(dt):
+        """ Tries to handle various date formats, not exhaustive but common, and extract the year"""
+
+        if isinstance(dt, dict):
+            year = dt.get('year')
+        elif re.match(r"^[0-9]{8}$", dt):  # 20201231
+            year = dt[0:4]
+        elif re.match(r"^[0-9]{6}$", dt):  # 201231
+            year = dt[0:4]
+        elif re.match(r"^[0-9]{4}$", dt):  # 2012
+            year = dt[0:4]
+        elif re.match(r"^[0-9]{4}[-/]", dt):  # 2020-12-31 or 2020/12/31
+            year = dt[0:4]
+        elif re.match(r"^[0-9]{2}/[0-9]{2}/[0-9]{4}", dt):  # 12/31/2020
+            year = dt[6:10]
+        else:
+            year = "0"
+
+        return int(year)
+
```

Comment:
We can do something like:

```python
from collections import Mapping
import re

YEAR_REGEX_YMD = re.compile(r'^(?P<year>\d{4})[-/]?(?:\d{2}[-/]?)?(?:\d{2}[-/]?)?$')
YEAR_REGEX_DMY = re.compile(r'^(?:\d{2}[-/])?(?:\d{2}[-/])?(?P<year>\d{4})$')


def extract_year_from_date(dt):
    """ Tries to handle various date formats, not exhaustive but common, and extract the year"""

    if isinstance(dt, Mapping):
        dt = dt.get('year', '')

    match = YEAR_REGEX_DMY.match(dt)
    if match:
        return int(match.group('year'))

    match = YEAR_REGEX_YMD.match(dt)
    if match:
        return int(match.group('year'))

    return 0


cases = (
    '20201231',
    '202031',
    '20203',
    '2020',
    '2020-12-31',
    '2020/12/31',
    '12/31/2020',
    '12-31-2020',
    '12-2020',
    '01-01-2020',
    'x',
    {'year': '2020'},
    {'year': 'x'},
    {'year': '2020/12/31'}
)

for dt in cases:
    print("%s -> %d" % (dt, extract_year_from_date(dt)))
```


Output:

```
20201231 -> 2020
202031 -> 2020
20203 -> 0
2020 -> 2020
2020-12-31 -> 2020
2020/12/31 -> 2020
12/31/2020 -> 2020
12-31-2020 -> 2020
12-2020 -> 2020
01-01-2020 -> 2020
x -> 0
{'year': '2020'} -> 2020
{'year': 'x'} -> 0
{'year': '2020/12/31'} -> 2020
```


Also, for such function, it is mandatory to write test code (in `test/`).
Also, this function should go in `picard/util` as it's a static method, not specific to `Metadata`.

Reply:
Agreed, especially putting this to util. Also some existing code in Picard assumes dates have the format `yyyy-mm-dd` and year can be extracted as the first four characters. Having this logic in utils allows us to review the cases and reuse the year extraction.